staging/src/k8s.io/api/core/v1/types.go

```diff
@@ -3524,22 +3524,31 @@ type ClaimSource struct {
 	//
 	// The template will be used to create a new ResourceClaim, which will
 	// be bound to this pod. When this pod is deleted, the ResourceClaim
-	// will also be deleted. The name of the ResourceClaim will be <pod
-	// name>-<resource name>, where <resource name> is the
-	// PodResourceClaim.Name. Pod validation will reject the pod if the
-	// concatenated name is not valid for a ResourceClaim (e.g. too long).
-	//
-	// An existing ResourceClaim with that name that is not owned by the
-	// pod will not be used for the pod to avoid using an unrelated
-	// resource by mistake. Scheduling and pod startup are then blocked
-	// until the unrelated ResourceClaim is removed.
+	// will also be deleted. The ResourceClaim will have a name that is
+	// generated with <pod name>-<resource name> as base and the Pod as
+	// owner. The generated name gets recorded in
+	// pod.status.resourceClaimStatuses.
 	//
 	// This field is immutable and no changes will be made to the
 	// corresponding ResourceClaim by the control plane after creating the
 	// ResourceClaim.
 	ResourceClaimTemplateName *string `json:"resourceClaimTemplateName,omitempty" protobuf:"bytes,2,opt,name=resourceClaimTemplateName"`
 }
 
+// PodResourceClaimStatus is stored in the PodStatus for each PodResourceClaim
+// which references a ResourceClaimTemplate. It stores the generated name for
+// the corresponding ResourceClaim.
+type PodResourceClaimStatus struct {
+	// Name uniquely identifies this resource claim inside the pod.
+	// This must match the name of an entry in pod.spec.resourceClaims,
+	// which implies that the string must be a DNS_LABEL.
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
```

Comment:
See https://github.com/kubernetes/kubernetes/pull/117351/commits/5d8727241d9dec2def88648223823b2901a28f4b for a revision of this API which should work for ResourceClaims that were generated for a network.

Reply:
I wasn't happy with https://github.com/kubernetes/kubernetes/commit/5d8727241d9dec2def88648223823b2901a28f4b. If we ever do something for networks, let's add a `pod.status.networkStatuses` that code can check to find a generated ResourceClaim.

The part about better error handling in `resourceclaim.Name` callers and allowing ResourceClaim creation to be skipped still made sense to me, so I kept that in the latest revision of this PR, after squashing and rebasing (because of a conflict).